frameworks/hdfs/src/main/java/com/mesosphere/sdk/hdfs/scheduler/HdfsRecoveryPlanManager.java

```diff
@@ -0,0 +1,167 @@
+        updatePlan(dirtyAssets);
+
+        Map<PodInstance, List<Protos.TaskInfo>> failedNameNodes;
+        try {
+            failedNameNodes = TaskUtils.getPodMap(
+                    configStore,
+                    StateStoreUtils.fetchTasksNeedingRecovery(stateStore, configStore));
+        } catch (TaskException e) {
+            logger.error("Failed to get tasks needing recovery, falling back to default recovery plan manager.", e);
+            return super.getCandidates(dirtyAssets);
+        }
+
+        failedNameNodes = failedNameNodes.entrySet().stream()
+                .filter(podInstanceListEntry -> podInstanceListEntry.getKey().getPod().getType().equals("name"))
+                .collect(Collectors.toMap(
+                        podInstanceListEntry -> podInstanceListEntry.getKey(),
+                        podInstanceListEntry -> podInstanceListEntry.getValue()));
+
+        Predicate<Protos.TaskInfo> isPodPermanentlyFailed = t -> (
+                FailureUtils.isLabeledAsFailed(t) || failureMonitor.hasFailed(t));
+
+        for (Map.Entry<PodInstance, List<Protos.TaskInfo>> failedPod : failedNameNodes.entrySet()) {
+            PodInstance podInstance = failedPod.getKey();
+            List<Protos.TaskInfo> failedTasks = failedPod.getValue();
+            if (failedTasks.stream().allMatch(isPodPermanentlyFailed) && nameNodeRecoveryPhase.isComplete()) {
+                switch (podInstance.getIndex()) {
+                    case 0:
+                        logger.info("Returning replacement plan for namenode 0.");
+                        setNameNodePhase(initNNRecoveryPhase(replaceNN0Plan));
+                        break;
+                    case 1:
+                        logger.info("Returning replacement plan for namenode 1.");
+                        setNameNodePhase(initNNRecoveryPhase(replaceNN1Plan));
+                        break;
+                    default:
+                        logger.error(
+                                "Encountered unexpected index: {}, falling back to default recovery plan manager.",
+                                podInstance.getIndex());
+                }
+            }
+        }
+
+        return PlanUtils.getCandidates(getPlan(), dirtyAssets).stream()
+                .filter(step ->
+                        launchConstrainer.canLaunch(((DefaultRecoveryStep) step).getRecoveryType()))
+                .collect(Collectors.toList());
+    }
+
+    private void setNameNodePhase(Phase phase) {
+        this.nameNodeRecoveryPhase = phase;
+    }
+
+    private Phase initNNRecoveryPhase(Plan inputPlan) {
+        Phase inputPhase = inputPlan.getChildren().get(0);
+        Step inputBootStrapStep = inputPhase.getChildren().get(0);
+        Step bootstrapStep =
+                new DefaultRecoveryStep(
+                        inputBootStrapStep.getName(),
+                        Status.PENDING,
+                        inputBootStrapStep.start().get().getPodInstance(),
```

Comment:
We're not recapitulating its structure.  Just converting from Deploy to Recovery steps.  Much easier than constructing plan from Java entirely.